packages/wdio-shadow-dom-service/tests/service.test.js

```diff
@@ -0,0 +1,89 @@
+import ShadowDomService from '../src'
+
+describe('ShadowDomService', () => {
+
+    let service
+    beforeAll(() => {
+        service = new ShadowDomService()
+        global.browser = {addCommand: jest.fn()}
+    })
+    describe('before', () => {
+        it('should add the commands', () => {
+            service.before()
+            expect(global.browser.addCommand.mock.calls).toHaveLength(2)
+            expect(global.browser.addCommand.mock.calls[0][0]).toBe('shadow$')
+            expect(global.browser.addCommand.mock.calls[0][1]).toEqual(expect.any(Function))
+            expect(global.browser.addCommand.mock.calls[0][2]).toBe(true)
+            expect(global.browser.addCommand.mock.calls[1][0]).toBe('shadow$$')
+            expect(global.browser.addCommand.mock.calls[1][1]).toEqual(expect.any(Function))
+            expect(global.browser.addCommand.mock.calls[1][2]).toBe(true)
+        })
+    })
+
+    describe('shadow$', () => {
+        let client
+        beforeAll(() => {
+            client = {$: jest.fn()}
+        })
+        it('should call client.$ with a function', () => {
+            ShadowDomService.shadow$.call(client, '#foo')
+            expect(client.$.mock.calls).toHaveLength(1)
+            expect(client.$.mock.calls[0]).toEqual([expect.any(Function)])
+        })
+    })
+
+    describe('shadow$$', () => {
```

Comment:
Take a look at the $ and $$() unit tests. We have a whole mocked out request system you can use. We need to ensure that the correct endpoint is actually hit.

Based on the selector you pass in, we can determine which mocked out response you get.

Reply:
Is that necessary? The correct endpoints would be whatever $ and $$ use. These commands just call $ or $$ with a function selector.